tsl/src/remote/txn.c

```diff
@@ -91,19 +94,24 @@ remote_txn_begin(RemoteTxn *entry, int curlevel)
 			sql = "START TRANSACTION ISOLATION LEVEL SERIALIZABLE";
 		else
 			sql = "START TRANSACTION ISOLATION LEVEL REPEATABLE READ";
+
+		remote_connection_xact_transition_begin(entry->conn);
```

Comment:
The functionality is not equivalent. Looking at the original postgres_fdw, on which our code is base, it seems functionality that was originally there was removed.

The deferred error thing only covered errors in subtransactions (actually only used in one single place, AFAIK). The guards (re-)introduced in this PR actually marks whether every command that involves a transaction state change has fully completed. If that's the case at the end of the transaction, the connection is OK to keep in the cache, otherwise it has to be closed.

This means that connections can be kept, as long as errors don't happen in any transaction state change command (around which the guards exist). The old code simply threw away the complete cache on every rollback irrespective of whether this was a user-initiated rollback or a rollback as a result of an error (e.g., unique constraint violation).